File: leetcode/baniry_tree/124.cpp
class Solution {
public:
    int res = INT32_MIN;

    int helper(TreeNode *root) {
        if (!root)
            return 0;

        int left = helper(root->left);
        int right = helper(root->right);
        res = max({left + root->val, right + root->val, left + right + root->val, root->val, res});
        return max({left + root->val, right + root->val, root->val});
    }

    int maxPathSum(TreeNode *root) {
        int x = helper(root);
        return res > x ? res : x;
    }
};

File: leetcode/baniry_tree/124.go
var res int

func findMax(args ...int) int {
	res := args[0]

	for _, arg := range args{
		if arg > res {
			res = arg
		}
	}
	
	return res
}

func helper(root *TreeNode) int {
	if root == nil {
		return 0
	}
	l := helper(root.Left)
	r := helper(root.Right)
	res = findMax(l + root.Val, r + root.Val, r + l + root.Val, root.Val, res)
	return findMax(l + root.Val, r + root.Val, root.Val)
}

func maxPathSum(root *TreeNode) int {
    res = math.MinInt32
	helper(root)
	return res
}